### PR TITLE
add features to slope charts

### PR DESCRIPTION
**Pull request by sherinkurian**
On slope charts:

- [x] Differentiate between hover and focus states
- [x] Add legend
- [x] Add search
- [x] Add foreground and background groups

As requested in:
https://github.com/owid/owid-grapher/issues/50
https://github.com/owid/owid-grapher/issues/51

@mispy

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -19,3 +19,4 @@ dist
 public
 grapher_admin/static/build
 .netlify
+*.swp
```
Comment (sherinkurian) [inline .gitignore]: for vim

File: js/charts/SlopeChart.tsx [hunk 2]
```diff
@@ -1,25 +1,119 @@
...
 
         const { bounds, chart } = this.props
         const { yAxis } = chart
         const { data } = this.transform
+        const { legend, focusKeys, hoverKeys, focusColors, activeColors, sidebarWidth } = this
 
-        return <LabelledSlopes bounds={bounds} yDomain={yAxis.domain} yTickFormat={this.transform.yTickFormat} yScaleType={yAxis.scaleType} yScaleTypeOptions={yAxis.scaleTypeOptions} onScaleTypeChange={(scaleType) => { chart.yAxis.scaleType = scaleType }} data={data} fontSize={chart.baseFontSize} />
+        return <g>
+            <LabelledSlopes bounds={bounds} yDomain={yAxis.domain} yTickFormat={this.transform.yTickFormat} yScaleType={yAxis.scaleType} yScaleTypeOptions={yAxis.scaleTypeOptions} onScaleTypeChange={(scaleType) => { chart.yAxis.scaleType = scaleType }} data={data} fontSize={chart.baseFontSize} focusKeys={hoverKeys} onMouseOver={this.onSlopeMouseOver} onMouseLeave={this.onSlopeMouseLeave}  />
+            <ScatterColorLegendView legend={legend} x={bounds.right - sidebarWidth} y={bounds.top} onMouseOver={this.onLegendMouseOver} onMouseLeave={this.onLegendMouseLeave} focusColors={focusColors} activeColors={activeColors} />
```
Comment (sherinkurian) [inline js/charts/SlopeChart.tsx]: may be rename `ScatterColorLegendView` to something generic?

**Comment by mispy** [comment]
Hi Sherin! Thanks so much for your help with this!

The group hover highlighting works perfectly. However, the bounds calculation is a bit off; the legend can overlap with the chart like this:

screen shot 2018-06-06 at 4 51 46 pm

To fix this I think you would need to correct the bounds for the slope chart based on the bounds of the legend (in the way the scatterplot also does)

**Comment by sherinkurian** [comment]
Hi Mispy, thanks for the review :) - the overlap is corrected.

minor thing - I felt the on-mouse-over experience for slope lines were a bit jumpy. but that could be because of too many data points.

**Comment by mispy** [comment]
Looking good!

It's true that the mouseover can be a bit jumpy. We might not want to reduce the area that much though-- it can be tricky to get the mouse exactly on very thin lines. Might be good to have an upper bound on distance, so that it activates the closest line but only if it is within a certain range.

There are a couple of compatibility things I need to work out with existing charts before we can merge this. I'll see if I can write a better export script so I can get you a more extensive copy of our database for testing with. :)

**Comment by mispy** [comment]
Hi Sherin, for this one I think the only blocker is compatibility with existing slope charts. I've ported the testing pages to the node admin on master now, and in conjunction with the database exports, you should hopefully be able to test how your code renders the charts in comparison to the live version by going here: http://localhost:3030/admin/test/embeds?type=SlopeChart

**Comment by sherinkurian** [comment]
Woah, this is great. I should be able to test this once the static build completes and all the urls start loading. Thank you for improving the development experience 💯

**Comment by sherinkurian** [comment]
Hi Jaiden, the features should be backwards compatible now. Thank you for the test pages. They were super helpful.

**Comment by mispy** [comment]
All works perfectly! Merged and deployed 🎉https://ourworldindata.org/grapher/child-mortality-by-country

**Comment by sherinkurian** [comment]
yay, that makes me so happy :)

**Pull request merged**